### ISSUE TITLE
`updatespendtx`: merge signatures, don't overwrite the PSBT entirely

### ISSUE THREAD

**Issue by darosior** [revault/revaultd#312]
It would greatly improve the UX in the GUI (because it wouldn't discard existing signatures).

**Comment by darosior** [comment]
Some clarifications about this. The `updatespendtx` RPC takes a PSBT of a Spend transaction to update the one in DB. If there is none in DB it stores it, otherwise it overwrites it.

Instead of overwriting it, it would be preferable to query the in-DB PSBT, add to it the signatures from the new PSBT and then store it. Make sure to use `revault_tx`'s `add_signature()` API which will check the validity of the signatures.

**Comment by Zshan0** [comment]
Hey, To update the progress, I'll list out the steps that I planned on taking
- Store the transaction that is being fetched by calling the function `db_spend_transaction`.
- If it exists then merge the signatures
- Pass in the new transaction to `db_update_spend` which is already being done

Although the suggested way was to call `add_signature()`, after a little exploration I found `db_txs_merge_sigs` and `revault_txs_merge_sigs`  present in `database/actions.rs`, I can't directly use them because 
1. The functions take transactions of the same type. And from what I understand, we will have the parameter transaction of type `SpendTransaction` whereas the transaction fetched is of the type `DbSpendTransaction`.
2. Both the functions are private.

I haven't checked if there is any way of typecasting `SpendTransaction` to `DbSpendTransaction`, if there is please let me know. In my opinion, we should use the pre-existing functionality to merge the signatures of two transactions rather than writing the functionality again in `update_spend_tx`.

**Comment by darosior** [comment]
> the steps that I planned on taking

Sounds good to me.

> Although the suggested way was to call add_signature(), after a little exploration  I found db_txs_merge_sigs and revault_txs_merge_sigs [...]   In my opinion, we should use the pre-existing functionality to merge the signatures of two transactions rather than  [...]

First of all, note how those two functions you talk about trust their inputs. For instance:
https://github.com/revault/revaultd/blob/457ff85dc40834ccf9e78ced25547a77d7278ad5/src/database/actions.rs#L769
That's because they are internal helpers working with data internal to the DB. Using such functions with inputs from the RPC would mean that we'd have to sanitize inputs first, and the easiest way of doing that is to use the `add_signature` API from `revault_tx`. Therefore we'd just be doing the work twice.

Secondly, those helpers are specific to presigned transactions (Unvault, Cancel and bother Emergency) which are managed separately from the Spend transactions in database:
https://github.com/revault/revaultd/blob/457ff85dc40834ccf9e78ced25547a77d7278ad5/src/database/schema.rs#L74-L123

Lastly, i'd like to get rid of these ugly hacks that `db_txs_merge_sigs` and friends are. But that's part of a big undertaking (#368).

**Comment by Zshan0** [comment]
Hey, continuing on this issue, I encountered more problem. Before I enumerate them, I thought I'll go through the process. First, I looked at the source code of `set_spend_tx` for inspiration and was able to get the list of signatures present in the transaction that is being fetched from the database, namely `old_tx`. 

```rs
let signatures: Vec<BTreeMap<BitcoinPubKey, Vec<u8>>> = old_tx
    .unwrap()
    .psbt
    .psbt()
    .inputs
    .iter()
    .map(|i| i.partial_sigs.clone())
    .collect();
```
Then I iterate over them and then add the signature to `spend_tx` transaction. After that, I will pass the newly updated transaction to `db_update_spend`. In `set_spend_tx`, it expects `"the signature was already there"` error while adding signatures. If there is an overlap of signatures, this will remain how it is.

But do I have to also any other expected error that the system could encounter while adding the new signatures?

If it's more convenient, I can create a PR and then there will be more context to what I'm referring to.

**Comment by darosior** [comment]
I think `set_spend_tx` is a good source of inspiration, what we do there to sanity check the sigs is what we should do here to add them:
https://github.com/revault/revaultd/blob/457ff85dc40834ccf9e78ced25547a77d7278ad5/src/commands/mod.rs#L1245-L1247

> In set_spend_tx, it expects "the signature was already there" error while adding signatures

It doesn't expect an error, it asserts that [`add_signature`](https://docs.rs/revault_tx/0.4.1/revault_tx/transactions/trait.RevaultTransaction.html#tymethod.add_signature) did not return `None`. It must never do so because the signature was already there when we added it.

**Comment by Zshan0** [comment]
Since I'm going to use `add_signature` to merge, I will have to call it on either `old_tx` or `spend_tx` to merge. I believe that I can't do so on the one that has been fetched from the database, but it is possible on `spend_tx` which has been passed as the argument. This would require it to be mutable. This will require changes in `jsonrpc/api.rs`. I don't see elsewhere in the API where `&mut` is being used So I'm not sure if this is the right direction, Is this change okay? or is there some other alternative that I can take?

https://github.com/revault/revaultd/blob/b310208a8b3b420f9b0cddcc03588e264a99538b/src/jsonrpc/api.rs#L441-L448

**Comment by darosior** [comment]
> I believe that I can't do so on the one that has been fetched from the database

Why wouldn't you?

> is there some other alternative that I can take?

[As previously stated](https://github.com/revault/revaultd/issues/312#issuecomment-1079665163), i think it should be done on the PSBT from the DB. It just sounds "safer".

**Comment by darosior** [comment]
This was fixed by #391.
